Some tests are failing on latest Node v22.2.0

On https://github.com/eslint/create-config/pull/115, I got several unrelated tests failing on Node v22.2.0.

Last time we ran tests on Node 22, they were passing, but that was with Node v22.1.0.

I re-ran Node 22 test on the main branch to check again with the latest Node v22.2.0, and it's failing.

https://github.com/eslint/create-config/actions/runs/9064687625/job/25110939311

```
  ❯ tests/utils/npm-utils.spec.js  (23 tests | 4 failed) 64ms
    ❯ tests/utils/npm-utils.spec.js > npmUtils > checkDevDeps() > should throw with message when parsing invalid package.json
      → expected [Function] to throw an error
    ❯ tests/utils/npm-utils.spec.js > npmUtils > checkDeps() > should throw with message when parsing invalid package.json
      → expected [Function] to throw an error
    ❯ tests/utils/npm-utils.spec.js > npmUtils > checkPackageJson() > should return false if package.json does not exist
      → expected true to equal false
    ❯ tests/utils/npm-utils.spec.js > npmUtils > installSyncSaveDev() > should log an error message if npm throws ENOENT error
      → Unspecified AssertionError
 stderr | tests/utils/npm-utils.spec.js > npmUtils > installSyncSaveDev() > should log an error message if npm throws ENOENT error
 Could not execute npm. Please install the following package with a package manager of your choice: some-package

 (node:1814) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
 --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("esmock", pathToFileURL("./"));'
 (Use `node --trace-warnings ...` to show where the warning was created)
  ✓ tests/config-snapshots.spec.js  (39 tests) 2420ms

 ⎯⎯⎯⎯⎯⎯⎯ Failed Tests 4 ⎯⎯⎯⎯⎯⎯⎯

  FAIL  tests/utils/npm-utils.spec.js > npmUtils > checkDevDeps() > should throw with message when parsing invalid package.json
 AssertionError: expected [Function] to throw an error

 - Expected:
 null

 + Received:
 undefined

  ❯ tests/utils/npm-utils.spec.js:84:20
      82|             });
      83|
      84|             assert.throws(() => {
        |                    ^
      85|                 stubcheckDevDeps(["some-package"]);
      86|             }, /JSON/u);


 ##[error]AssertionError: expected [Function] to throw an error

    - Expected:
    null

    + Received:
    undefined

     ❯ tests/utils/npm-utils.spec.js:84:20


 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/4]⎯

  FAIL  tests/utils/npm-utils.spec.js > npmUtils > checkDeps() > should throw with message when parsing invalid package.json
 AssertionError: expected [Function] to throw an error

 - Expected:
 null

 + Received:
 undefined

  ❯ tests/utils/npm-utils.spec.js:134:20

 ##[error]AssertionError: expected [Function] to throw an error

    - Expected:
    null

    + Received:
    undefined

     ❯ tests/utils/npm-utils.spec.js:134:20


     132|             });
     133|
     134|             assert.throws(() => {
        |                    ^
     135|                 stubbedcheckDeps(["some-package"]);
     136|             }, /JSON/u);

 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/4]⎯

  FAIL  tests/utils/npm-utils.spec.js > npmUtils > checkPackageJson() > should return false if package.json does not exist
 AssertionError: expected true to equal false

 - Expected
 + Received

 - false
 + true

  ❯ tests/utils/npm-utils.spec.js:152:20

 ##[error]AssertionError: expected true to equal false

    - Expected
    + Received

    - false
    + true

     ❯ tests/utils/npm-utils.spec.js:152:20


     150|             const { checkPackageJson: stubbedcheckPackageJson } = awai…

 ##[error]AssertionError: Unspecified AssertionError
     ❯ tests/utils/npm-utils.spec.js:200:13


     151|
     152|             assert.strictEqual(stubbedcheckPackageJson(), false);
        |                    ^
     153|         });
     154|     });

 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[3/4]⎯

  FAIL  tests/utils/npm-utils.spec.js > npmUtils > installSyncSaveDev() > should log an error message if npm throws ENOENT error
 AssertionError: Unspecified AssertionError
  ❯ tests/utils/npm-utils.spec.js:200:13
     198|             stubinstallSyncSaveDev("some-package");
     199|
     200|             assert(logErrorStub.calledOnce);
        |             ^
     201|
     202|             npmUtilsStub.restore();

 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[4/4]⎯

  Test Files  1 failed | 1 passed (2)
       Tests  4 failed | 58 passed (62)
    Start at  16:45:59
    Duration  2.85s (transform 89ms, setup 0ms, collect 239ms, tests 2.48s, environment 1ms, prepare 281ms)

```

--- a/.npmrc
+++ b/.npmrc
@@ -1,2 +1,1 @@
 package-lock=false
-node-options=--loader=esmock

--- a/package.json
+++ b/package.json
@@ -48,7 +48,6 @@
         "eslint": "^9.1.1",
         "eslint-config-eslint": "^10.0.0",
         "eslint-release": "^3.2.0",
-        "esmock": "^2.5.8",
         "lint-staged": "^12.1.2",
         "memfs": "^3.4.0",
         "sinon": "^12.0.1",

--- a/tests/utils/npm-utils.spec.js
+++ b/tests/utils/npm-utils.spec.js
@@ -13,10 +13,10 @@ import {
     installSyncSaveDev,
     fetchPeerDependencies,
     checkDeps,
-    checkDevDeps
+    checkDevDeps,
+    checkPackageJson
 } from "../../lib/utils/npm-utils.js";
 import { defineInMemoryFs } from "../_utils/in-memory-fs.js";
-import esmock from "esmock";
 import { assert, describe, afterEach, it } from "vitest";
 
 //------------------------------------------------------------------------------
@@ -24,14 +24,17 @@ import { assert, describe, afterEach, it } from "vitest";
 //------------------------------------------------------------------------------
 
 /**
- * Import `npm-utils` with the in-memory file system.
+ * Replace native fs methods with the in-memory file system methods used by `npm-utils`.
  * @param {Object} files The file definitions.
- * @returns {Object} `npm-utils`.
+ * @returns {void}
  */
-async function requireNpmUtilsWithInMemoryFileSystem(files) {
-    const fs = defineInMemoryFs({ files });
+async function useInMemoryFileSystem(files) {
+    const inMemoryFs = defineInMemoryFs({ files });
+    const { default: fs } = await import("fs");
 
-    return await esmock("../../lib/utils/npm-utils.js", { fs });
+    sinon.replace(fs, "readFileSync", inMemoryFs.readFileSync);
+    sinon.replace(fs, "existsSync", inMemoryFs.existsSync);
+    sinon.replace(fs, "statSync", inMemoryFs.statSync);
 }
 
 //------------------------------------------------------------------------------
@@ -68,21 +71,21 @@ describe("npmUtils", () => {
         });
 
         it("should handle missing devDependencies key", async () => {
-            const { checkDevDeps: stubcheckDevDeps } = await requireNpmUtilsWithInMemoryFileSystem({
+            await useInMemoryFileSystem({
                 "package.json": JSON.stringify({ private: true, dependencies: {} })
             });
 
             // Should not throw.
-            stubcheckDevDeps(["some-package"]);
+            checkDevDeps(["some-package"]);
         });
 
         it("should throw with message when parsing invalid package.json", async () => {
-            const { checkDevDeps: stubcheckDevDeps } = await requireNpmUtilsWithInMemoryFileSystem({
+            await useInMemoryFileSystem({
                 "package.json": '{ "not: "valid json" }'
             });
 
             assert.throws(() => {
-                stubcheckDevDeps(["some-package"]);
+                checkDevDeps(["some-package"]);
             }, /JSON/u);
         });
     });
@@ -118,38 +121,38 @@ describe("npmUtils", () => {
         });
 
         it("should handle missing dependencies key", async () => {
-            const { checkDeps: stubbedcheckDeps } = await requireNpmUtilsWithInMemoryFileSystem({
+            await useInMemoryFileSystem({
                 "package.json": JSON.stringify({ private: true, devDependencies: {} })
             });
 
             // Should not throw.
-            stubbedcheckDeps(["some-package"]);
+            checkDeps(["some-package"]);
         });
 
         it("should throw with message when parsing invalid package.json", async () => {
-            const { checkDeps: stubbedcheckDeps } = await requireNpmUtilsWithInMemoryFileSystem({
+            await useInMemoryFileSystem({
                 "package.json": '{ "not: "valid json" }'
             });
 
             assert.throws(() => {
-                stubbedcheckDeps(["some-package"]);
+                checkDeps(["some-package"]);
             }, /JSON/u);
         });
     });
 
     describe("checkPackageJson()", () => {
         it("should return true if package.json exists", async () => {
-            const { checkPackageJson: stubbedcheckPackageJson } = await requireNpmUtilsWithInMemoryFileSystem({
+            await useInMemoryFileSystem({
                 "package.json": '{ "file": "contents" }'
             });
 
-            assert.strictEqual(stubbedcheckPackageJson(), true);
+            assert.strictEqual(checkPackageJson(), true);
         });
 
         it("should return false if package.json does not exist", async () => {
-            const { checkPackageJson: stubbedcheckPackageJson } = await requireNpmUtilsWithInMemoryFileSystem({});
+            await useInMemoryFileSystem({});
 
-            assert.strictEqual(stubbedcheckPackageJson(), false);
+            assert.strictEqual(checkPackageJson(), false);
         });
     });
 
@@ -188,14 +191,11 @@ describe("npmUtils", () => {
         it("should log an error message if npm throws ENOENT error", async () => {
             const logErrorStub = sinon.spy();
             const npmUtilsStub = sinon.stub(spawn, "sync").returns({ error: { code: "ENOENT" } });
+            const log = await import("../../lib/utils/logging.js");
 
-            const { installSyncSaveDev: stubinstallSyncSaveDev } = await esmock("../../lib/utils/npm-utils.js", {
-                "../../lib/utils/logging.js": {
-                    error: logErrorStub
-                }
-            });
+            sinon.replaceGetter(log, "error", () => logErrorStub);
 
-            stubinstallSyncSaveDev("some-package");
+            installSyncSaveDev("some-package");
 
             assert(logErrorStub.calledOnce);
 

--- a/tests/utils/npm-utils.spec.js
+++ b/tests/utils/npm-utils.spec.js
@@ -18,6 +18,7 @@ import {
 } from "../../lib/utils/npm-utils.js";
 import { defineInMemoryFs } from "../_utils/in-memory-fs.js";
 import { assert, describe, afterEach, it } from "vitest";
+import fs from "fs";
 
 //------------------------------------------------------------------------------
 // Helpers
@@ -30,7 +31,6 @@ import { assert, describe, afterEach, it } from "vitest";
  */
 async function useInMemoryFileSystem(files) {
     const inMemoryFs = defineInMemoryFs({ files });
-    const { default: fs } = await import("fs");
 
     sinon.replace(fs, "readFileSync", inMemoryFs.readFileSync);
     sinon.replace(fs, "existsSync", inMemoryFs.existsSync);

